File: XMUOJ/FTCT-47/main.py
input()
input()
line1 = input().strip()
line2 = input().strip()
line3 = input().strip()
input()
input()


def for_loop(string) -> list:
    name = string.split()[0].strip()
    if "step" in string:
        step = int(string.split("step")[-1].strip())
    else:
        step = 1
    if "downTo" in string:
        start = int(string.split("downTo")[0].split()[-1].strip())
        end = int(string.split("downTo")[-1].split()[0].strip())
        step = -step
    else:
        start = int(string.split("..")[0].split()[-1].strip())
        end = int(string.split("..")[-1].split()[0].strip())
    return [name, start, end, step]


def get_range(loop: list) -> list:
    if loop[1] == loop[2]:
        return [loop[1]]
    if loop[2] > 0:
        loop_range = range(loop[1], loop[2] + 1, loop[3])
    elif loop[2] < 0:
        loop_range = range(loop[1], loop[2] - 1, loop[3])
    else:
        if loop[1] > 0:
            loop_range = range(loop[1], loop[2] - 1, loop[3])
        else:
            loop_range = range(loop[1], loop[2] + 1, loop[3])
    return loop_range


def get_sum(loop : list) -> int:
    length = len(loop)
    return (loop[0] + loop[-1]) * length // 2


outer = for_loop(line1.split("(")[-1].split(")")[0].strip())
inner = for_loop(line2.split("(")[-1].split(")")[0].strip())
add = line3.split("+=")[-1].strip()

res = 0
outer_range = get_range(outer)
inner_range = get_range(inner)

if outer[0] == add:
    res += get_sum(outer_range) * len(inner_range)
else:
    res += get_sum(inner_range) * len(outer_range)
print(res)

File: XMUOJ/FTCT-47/main.cpp
#include <bits/stdc++.h>
using namespace std;

struct LoopInfo
{
    string name;
    long long sum;
    long long length;
};


vector<string> split(const string &s, char delim)
{
    vector<string> elems;
    stringstream ss(s);
    string item;
    while (getline(ss, item, delim))
    {
        elems.push_back(item);
    }
    return elems;
}

string trim(const string &s)
{
    string result = s;
    result.erase(0, result.find_first_not_of(" "));
    result.erase(result.find_last_not_of(" ") + 1);
    return result;
}

LoopInfo resolve_for_loop(string s)
{
    vector<string> parts = split(s, ' ');
    string name = trim(parts[0]);
    long long start, end, finalEnd;
    long long step = 1, length = 0, sum = 0;
    if (s.find("step") != string::npos)
    {
        step = stoi(trim(parts.back()));
        parts.pop_back();
        parts.pop_back();
    }
    if (s.find("downTo") != string::npos)
    {
        start = stoi(parts[2]);
        end = stoi(parts.back());
        step = -step;
    }
    else
    {
        vector<string> range = split(parts[2], '.');
        for (int i = 0; i < range.size(); i++)
        start = stoi(range[0]);
        end = stoi(range[2]);
    }
    if (start <= end)
        for (int i = start; i <= end; i += step)
        {
            finalEnd = i;
            length++;
        }
    else
        for (int i = start; i >= end; i += step)
        {
            finalEnd = i;
            length++;
        }
    sum = (start + finalEnd) * length / 2;
    return LoopInfo{name, sum, length};
}

int main()
{
    string tmp, line1, line2, line3;
    getline(cin, tmp);
    getline(cin, tmp);
    getline(cin, line1);
    getline(cin, line2);
    getline(cin, line3);
    getline(cin, tmp);
    getline(cin, tmp);

    LoopInfo outer = resolve_for_loop(split(split(line1, '(').back(), ')').front());
    LoopInfo inner = resolve_for_loop(split(split(line2, '(').back(), ')').front());
    string add = trim(split(line3, '=').back());

    int res;
    if (outer.name == add)
        res = outer.sum * inner.length;
    else
        res = inner.sum * outer.length;
    cout << res << endl;

    return 0;
}